File: shortest_palindrome.rb
# @param {String} s
# @return {String}
def shortest_palindrome(s)
  max = s.length - 1
  pointer = s.length
  s_ary = s.split('')
  parsed_chars = []
  l = 0
  s_ary.each_with_index do |x, idx|
    parsed_chars << x
    palin_finder = parsed_chars.join('')
    if palin_finder == palin_finder.reverse && palin_finder.length != 1
      l = palin_finder.length 
    end
  end

  ans = if l == 0
          (parsed_chars.drop(1).reverse + s_ary).join('')
        else
          (parsed_chars.drop(l).reverse + s_ary).join('')
        end
  ans
end

File: shortest-palindrome.js
/**
 * @param {string} s
 * @return {string}
 */
var shortestPalindrome = function(s) {
  var shift = 0;
  var l = s.length;
  var reverse = s.split('').reverse().join('');

  // check here if it's already a plaindrome
  var palindrome = true;
  for (var i=0; i<l/2; i++) {
    if (s[i] !== s[l-i-1]) {
      palindrome = false;
    }
  }
  if (palindrome) return s;

  // The trick is to start with the reverse (since that by nature is the worst case)
  // And shift the original string once to the right to act like we've appended a character
  // Once you get a match, you know that section is a palindrome and the rest need to be appended
  for (var i=1; i < s.length; i++) {
    if (s.slice(0, l-i) === reverse.slice(i, l)) {
      shift = i;
      break;
    }
  }

  var remainder = reverse.slice(0, shift);
  return remainder + s;
};
